In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from matplotlib.colors import LinearSegmentedColormap
from matplotlib import colors as mcolors
from scipy.stats import linregress
from sklearn.ensemble import IsolationForest
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

from sklearn.cluster import KMeans
from tabulate import tabulate
from collections import Counter
%matplotlib inline


import plotly.graph_objects as go
from sklearn import preprocessing
from matplotlib import colors
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')
import sys
if not sys.warnoptions:
    warnings.simplefilter('ignore')
np.random.seed(42)

import datetime
from datetime import datetime
from datetime import timedelta
from datetime import date

import plotly.express as px
import plotly.graph_objects as go
import plotly.io as pio

In [3]:
df = pd.read_csv("C:/Users/HARDEEP/OneDrive/Docs/Automobile-Sales-Logistics/Data/Raw-Data/Auto Sales data.csv")

In [4]:
df

ORDERNUMBER  QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER    SALES  \
0           10107               30      95.70                2  2871.00   
1           10121               34      81.35                5  2765.90   
2           10134               41      94.74                2  3884.34   
3           10145               45      83.26                6  3746.70   
4           10168               36      96.66                1  3479.76   
...           ...              ...        ...              ...      ...   
2742        10350               20     112.22               15  2244.40   
2743        10373               29     137.19                1  3978.51   
2744        10386               43     125.99                4  5417.57   
2745        10397               34      62.24                1  2116.16   
2746        10414               47      65.52                9  3079.44   

       ORDERDATE  DAYS_SINCE_LASTORDER    STATUS  PRODUCTLINE  MSRP  \
0     24/02/2018                   828   Shipped  Motorcycles    95   
1     07/05/2018                   757   Shipped  Motorcycles    95   
2     01/07/2018                   703   Shipped  Motorcycles    95   
3     25/08/2018                   649   Shipped  Motorcycles    95   
4     28/10/2018                   586   Shipped  Motorcycles    95   
...          ...                   ...       ...          ...   ...   
2742  02/12/2019                  2924   Shipped        Ships    54   
2743  31/01/2020                  2865   Shipped        Ships    54   
2744  01/03/2020                  2836  Resolved        Ships    54   
2745  28/03/2020                  2810   Shipped        Ships    54   
2746  06/05/2020                  2772   On Hold        Ships    54   

     PRODUCTCODE             CUSTOMERNAME             PHONE  \
0       S10_1678        Land of Toys Inc.        2125557818   
1       S10_1678       Reims Collectables        26.47.1555   
2       S10_1678          Lyon Souveniers  +33 1 46 62 7555   
3       S10_1678        Toys4GrownUps.com        6265557265   
4       S10_1678     Technics Stores Inc.        6505556809   
...          ...                      ...               ...   
2742    S72_3212    Euro Shopping Channel    (91) 555 94 44   
2743    S72_3212  Oulu Toy Supplies, Inc.        981-443655   
2744    S72_3212    Euro Shopping Channel    (91) 555 94 44   
2745    S72_3212             Alpha Cognac        61.77.6555   
2746    S72_3212        Gifts4AllAges.com        6175559555   

                       ADDRESSLINE1        CITY POSTALCODE  COUNTRY  \
0           897 Long Airport Avenue         NYC      10022      USA   
1                59 rue de l'Abbaye       Reims      51100   France   
2     27 rue du Colonel Pierre Avia       Paris      75508   France   
3                78934 Hillside Dr.    Pasadena      90003      USA   
4                 9408 Furth Circle  Burlingame      94217      USA   
...                             ...         ...        ...      ...   
2742             C/ Moralzarzal, 86      Madrid      28034    Spain   
2743                    Torikatu 38        Oulu      90110  Finland   
2744             C/ Moralzarzal, 86      Madrid      28034    Spain   
2745          1 rue Alsace-Lorraine    Toulouse      31000   France   
2746             8616 Spinnaker Dr.      Boston      51003      USA   

     CONTACTLASTNAME CONTACTFIRSTNAME DEALSIZE  
0                 Yu             Kwai    Small  
1            Henriot             Paul    Small  
2           Da Cunha           Daniel   Medium  
3              Young            Julie   Medium  
4             Hirano             Juri   Medium  
...              ...              ...      ...  
2742          Freyre            Diego    Small  
2743       Koskitalo           Pirkko   Medium  
2744          Freyre            Diego   Medium  
2745          Roulet          Annette    Small  
2746         Yoshido             Juri   Medium  

[2747 rows x 20 columns]

In [5]:
df.head(10)

ORDERNUMBER  QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER    SALES  \
0        10107               30      95.70                2  2871.00   
1        10121               34      81.35                5  2765.90   
2        10134               41      94.74                2  3884.34   
3        10145               45      83.26                6  3746.70   
4        10168               36      96.66                1  3479.76   
5        10180               29      86.13                9  2497.77   
6        10188               48     114.84                1  5512.32   
7        10211               41     114.84               14  4708.44   
8        10223               37     107.18                1  3965.66   
9        10237               23     101.44                7  2333.12   

    ORDERDATE  DAYS_SINCE_LASTORDER   STATUS  PRODUCTLINE  MSRP PRODUCTCODE  \
0  24/02/2018                   828  Shipped  Motorcycles    95    S10_1678   
1  07/05/2018                   757  Shipped  Motorcycles    95    S10_1678   
2  01/07/2018                   703  Shipped  Motorcycles    95    S10_1678   
3  25/08/2018                   649  Shipped  Motorcycles    95    S10_1678   
4  28/10/2018                   586  Shipped  Motorcycles    95    S10_1678   
5  11/11/2018                   573  Shipped  Motorcycles    95    S10_1678   
6  18/11/2018                   567  Shipped  Motorcycles    95    S10_1678   
7  15/01/2019                   510  Shipped  Motorcycles    95    S10_1678   
8  20/02/2019                   475  Shipped  Motorcycles    95    S10_1678   
9  05/04/2019                   432  Shipped  Motorcycles    95    S10_1678   

                 CUSTOMERNAME             PHONE  \
0           Land of Toys Inc.        2125557818   
1          Reims Collectables        26.47.1555   
2             Lyon Souveniers  +33 1 46 62 7555   
3           Toys4GrownUps.com        6265557265   
4        Technics Stores Inc.        6505556809   
5    Daedalus Designs Imports        20.16.1555   
6                Herkku Gifts     +47 2267 3215   
7            Auto Canal Petit    (1) 47.55.6555   
8  Australian Collectors, Co.      03 9520 4555   
9             Vitachrome Inc.        2125551500   

                    ADDRESSLINE1        CITY POSTALCODE    COUNTRY  \
0        897 Long Airport Avenue         NYC      10022        USA   
1             59 rue de l'Abbaye       Reims      51100     France   
2  27 rue du Colonel Pierre Avia       Paris      75508     France   
3             78934 Hillside Dr.    Pasadena      90003        USA   
4              9408 Furth Circle  Burlingame      94217        USA   
5        184, chausse de Tournai       Lille      59000     France   
6    Drammen 121, PR 744 Sentrum      Bergen     N 5804     Norway   
7              25, rue Lauriston       Paris      75016     France   
8              636 St Kilda Road   Melbourne       3004  Australia   
9              2678 Kingston Rd.         NYC      10022        USA   

  CONTACTLASTNAME CONTACTFIRSTNAME DEALSIZE  
0              Yu             Kwai    Small  
1         Henriot             Paul    Small  
2        Da Cunha           Daniel   Medium  
3           Young            Julie   Medium  
4          Hirano             Juri   Medium  
5           Rance          Martine    Small  
6          Oeztan           Veysel   Medium  
7         Perrier        Dominique   Medium  
8        Ferguson            Peter   Medium  
9           Frick          Michael    Small

INITIAL EXPLORATORY DATA ANALYSIS (EDA)

In [6]:
print('Samples:', df.shape[0])
print('Features:', df.shape[1])

Samples: 2747
Features: 20


In [7]:
df.DEALSIZE.value_counts() 

DEALSIZE
Medium    1349
Small     1246
Large      152
Name: count, dtype: int64

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2747 entries, 0 to 2746
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   ORDERNUMBER           2747 non-null   int64  
 1   QUANTITYORDERED       2747 non-null   int64  
 2   PRICEEACH             2747 non-null   float64
 3   ORDERLINENUMBER       2747 non-null   int64  
 4   SALES                 2747 non-null   float64
 5   ORDERDATE             2747 non-null   object 
 6   DAYS_SINCE_LASTORDER  2747 non-null   int64  
 7   STATUS                2747 non-null   object 
 8   PRODUCTLINE           2747 non-null   object 
 9   MSRP                  2747 non-null   int64  
 10  PRODUCTCODE           2747 non-null   object 
 11  CUSTOMERNAME          2747 non-null   object 
 12  PHONE                 2747 non-null   object 
 13  ADDRESSLINE1          2747 non-null   object 
 14  CITY                  2747 non-null   object 
 15  POSTALCODE           

In [9]:
df.isnull().sum()  # Missing Values

ORDERNUMBER             0
QUANTITYORDERED         0
PRICEEACH               0
ORDERLINENUMBER         0
SALES                   0
ORDERDATE               0
DAYS_SINCE_LASTORDER    0
STATUS                  0
PRODUCTLINE             0
MSRP                    0
PRODUCTCODE             0
CUSTOMERNAME            0
PHONE                   0
ADDRESSLINE1            0
CITY                    0
POSTALCODE              0
COUNTRY                 0
CONTACTLASTNAME         0
CONTACTFIRSTNAME        0
DEALSIZE                0
dtype: int64

In [10]:
df.duplicated().sum()    # Duplicate values 

0

In [11]:
df.describe()

ORDERNUMBER  QUANTITYORDERED    PRICEEACH  ORDERLINENUMBER  \
count   2747.000000      2747.000000  2747.000000      2747.000000   
mean   10259.761558        35.103021   101.098952         6.491081   
std       91.877521         9.762135    42.042549         4.230544   
min    10100.000000         6.000000    26.880000         1.000000   
25%    10181.000000        27.000000    68.745000         3.000000   
50%    10264.000000        35.000000    95.550000         6.000000   
75%    10334.500000        43.000000   127.100000         9.000000   
max    10425.000000        97.000000   252.870000        18.000000   

              SALES  DAYS_SINCE_LASTORDER         MSRP  
count   2747.000000           2747.000000  2747.000000  
mean    3553.047583           1757.085912   100.691664  
std     1838.953901            819.280576    40.114802  
min      482.130000             42.000000    33.000000  
25%     2204.350000           1077.000000    68.000000  
50%     3184.800000           1761.000000    99.000000  
75%     4503.095000           2436.500000   124.000000  
max    14082.800000           3562.000000   214.000000

FINDINGS:

1. The dataset contains 20 columns, or features, which will be feature engineered later to keep and use the relevant ones.
2. The total number of values or rows is 2747 and columns is 20
3. There are no missing values
4. There are zero duplicate values
5. sales ranges from 483 approx to 14082
6. MSRP per unit ranges from 33 to 214.
7. Quantity ordered minimum from 6 units to maximum 97 units.


NEXT STEPS:

1. Feature engineering would be done.
2. Additional derived features would be taken in to consideration originated from given features.
3. Name columns would be concantenated to get full name in order to get rid of extra columns so as to save space.
4. Similarly address, city and country columns would be merge together to get a complete address named column.
5. No adjsutment needed for missing or duplicate values since there are none.

In [12]:
df['CONTACTLASTNAME'].value_counts()   # here contactlast and contactfirst name have two value_counts different numbers so chances are there are some same firstnames 
                                       # but have different surnames. so we will join these two together to get a unique name column.

CONTACTLASTNAME
Freyre      259
Nelson      204
Young       115
Frick        91
Yu           80
           ... 
Hardy        12
Tseng        11
Kuo          10
Cartrain      8
Schmitt       7
Name: count, Length: 76, dtype: int64

In [13]:
df[['ADDRESSLINE1', 'COUNTRY']].value_counts()

ADDRESSLINE1                  COUNTRY  
C/ Moralzarzal, 86            Spain        259
5677 Strong St.               USA          180
636 St Kilda Road             Australia     55
67, rue des Cinquante Otages  France        53
Fauntleroy Circus             UK            51
                                          ... 
5290 North Pendale Street     USA           10
Boulevard Tirou, 255          Belgium        8
16780 Pompton St.             USA            8
54, rue Royale                France         7
4097 Douglas Av.              USA            3
Name: count, Length: 89, dtype: int64

In [14]:
# df['CONTACTNAME'] = df['CONTACTLASTNAME'] + ", " + df['CONTACTFIRSTNAME']

df['CONTACTNAME'] = df['CONTACTFIRSTNAME'] + " " + df['CONTACTLASTNAME']

print(df[['CONTACTFIRSTNAME', 'CONTACTLASTNAME', 'CONTACTNAME']].head(20))

   CONTACTFIRSTNAME CONTACTLASTNAME        CONTACTNAME
0              Kwai              Yu            Kwai Yu
1              Paul         Henriot       Paul Henriot
2            Daniel        Da Cunha    Daniel Da Cunha
3             Julie           Young        Julie Young
4              Juri          Hirano        Juri Hirano
5           Martine           Rance      Martine Rance
6            Veysel          Oeztan      Veysel Oeztan
7         Dominique         Perrier  Dominique Perrier
8             Peter        Ferguson     Peter Ferguson
9           Michael           Frick      Michael Frick
10          William           Brown      William Brown
11            Julie            King         Julie King
12           Janine         Labrune     Janine Labrune
13            Marta       Hernandez    Marta Hernandez
14            Matti       Karttunen    Matti Karttunen
15            Jonas      Bergulfsen   Jonas Bergulfsen
16            Kyung              Yu           Kyung Yu
17        

In [15]:
df['CONTACTNAME'].unique()

array(['Kwai Yu', 'Paul Henriot', 'Daniel Da Cunha', 'Julie Young',
       'Juri Hirano', 'Martine Rance', 'Veysel Oeztan',
       'Dominique Perrier', 'Peter Ferguson', 'Michael Frick',
       'William Brown', 'Julie King', 'Janine Labrune', 'Marta Hernandez',
       'Matti Karttunen', 'Jonas Bergulfsen', 'Kyung Yu', 'Georg Pipps',
       'Adrian Huxley', 'Violeta Benitez', 'Elizabeth Devon',
       'Diego Freyre', 'Christina Berglund', 'Mart¡n Sommer',
       'Eric Natividad', 'Maria Hernandez', 'Tony Calaghan',
       'Francisca Cervantes', 'Mary Saveley', 'Yoshi Tannamuri',
       'Steve Thompson', 'Kyung Tseng', 'Leslie Murphy',
       'Akiko Shimamura', 'Paolo Accorti', 'Maria Larsson',
       'Daniel Tonini', 'Valarie Nelson', 'Valarie Young', 'Allen Nelson',
       "Anna O'Hara", 'Jean Fresnisre', 'Mory Kentary', 'Rosa Hernandez',
       'Carine Schmitt', 'Steve Frick', 'Jytte Petersen', 'Ann Brown',
       'Wing C Tam', 'Annette Roulet', 'Thomas Hardy', 'Eduardo Saavedra',
   

In [16]:
df['COMPLETEADDRESS'] = (
    df['ADDRESSLINE1'].astype(str) + ", " +
    df['CITY'].astype(str) + ", " +
    df['POSTALCODE'].astype(str) + ", " +
    df['COUNTRY'].astype(str)
)


In [17]:
df.groupby('PRODUCTCODE')['COUNTRY'].value_counts().head(10)

PRODUCTCODE  COUNTRY  
S10_1678     USA          10
             France        6
             Australia     2
             Norway        2
             Austria       1
             Finland       1
             Spain         1
             UK            1
S10_1949     USA          10
             Australia     4
Name: count, dtype: int64

In [18]:
# I will drop some of the columns which is not that useful initially, but can be checked later, specially customer personal info and these info has been combined
# in to one column for more convenience.

df.drop(columns = ['PHONE', 'ADDRESSLINE1', 'POSTALCODE', 'CONTACTLASTNAME', 'CONTACTFIRSTNAME', 'ORDERLINENUMBER'], inplace = True)

In [19]:
df.columns = df.columns.str.lower()

In [20]:
df.insert(loc = 9, column = 'totalmsrp', value = df['msrp'] * df['quantityordered'])

In [21]:
col = df.pop('productcode')

In [22]:
df.insert(loc = 13, column = 'productcode', value = col)

In [23]:
df.rename(columns = {'status': 'shipping_status'}, inplace = True)

In [24]:
# further I will drop few more columns to avoid redundancy since we derived couple of new features which are more interpretable than the existing ones.
# Below given columns were dropped

df.drop(columns = ['priceeach', 'quantityordered', 'msrp'], inplace = True) 
                                                                             

In [25]:
df['shipping_status'] = df['shipping_status'].apply(lambda x: 'Shipped' if x == 'Shipped' else 'Not Shipped')

In [26]:
df

ordernumber    sales   orderdate  days_since_lastorder shipping_status  \
0           10107  2871.00  24/02/2018                   828         Shipped   
1           10121  2765.90  07/05/2018                   757         Shipped   
2           10134  3884.34  01/07/2018                   703         Shipped   
3           10145  3746.70  25/08/2018                   649         Shipped   
4           10168  3479.76  28/10/2018                   586         Shipped   
...           ...      ...         ...                   ...             ...   
2742        10350  2244.40  02/12/2019                  2924         Shipped   
2743        10373  3978.51  31/01/2020                  2865         Shipped   
2744        10386  5417.57  01/03/2020                  2836     Not Shipped   
2745        10397  2116.16  28/03/2020                  2810         Shipped   
2746        10414  3079.44  06/05/2020                  2772     Not Shipped   

      productline  totalmsrp             customername        city  country  \
0     Motorcycles       2850        Land of Toys Inc.         NYC      USA   
1     Motorcycles       3230       Reims Collectables       Reims   France   
2     Motorcycles       3895          Lyon Souveniers       Paris   France   
3     Motorcycles       4275        Toys4GrownUps.com    Pasadena      USA   
4     Motorcycles       3420     Technics Stores Inc.  Burlingame      USA   
...           ...        ...                      ...         ...      ...   
2742        Ships       1080    Euro Shopping Channel      Madrid    Spain   
2743        Ships       1566  Oulu Toy Supplies, Inc.        Oulu  Finland   
2744        Ships       2322    Euro Shopping Channel      Madrid    Spain   
2745        Ships       1836             Alpha Cognac    Toulouse   France   
2746        Ships       2538        Gifts4AllAges.com      Boston      USA   

     productcode dealsize       contactname  \
0       S10_1678    Small           Kwai Yu   
1       S10_1678    Small      Paul Henriot   
2       S10_1678   Medium   Daniel Da Cunha   
3       S10_1678   Medium       Julie Young   
4       S10_1678   Medium       Juri Hirano   
...          ...      ...               ...   
2742    S72_3212    Small      Diego Freyre   
2743    S72_3212   Medium  Pirkko Koskitalo   
2744    S72_3212   Medium      Diego Freyre   
2745    S72_3212    Small    Annette Roulet   
2746    S72_3212   Medium      Juri Yoshido   

                                        completeaddress  
0              897 Long Airport Avenue, NYC, 10022, USA  
1              59 rue de l'Abbaye, Reims, 51100, France  
2     27 rue du Colonel Pierre Avia, Paris, 75508, F...  
3              78934 Hillside Dr., Pasadena, 90003, USA  
4             9408 Furth Circle, Burlingame, 94217, USA  
...                                                 ...  
2742           C/ Moralzarzal, 86, Madrid, 28034, Spain  
2743                  Torikatu 38, Oulu, 90110, Finland  
2744           C/ Moralzarzal, 86, Madrid, 28034, Spain  
2745     1 rue Alsace-Lorraine, Toulouse, 31000, France  
2746             8616 Spinnaker Dr., Boston, 51003, USA  

[2747 rows x 14 columns]

In [27]:
data = df

In [28]:
data

ordernumber    sales   orderdate  days_since_lastorder shipping_status  \
0           10107  2871.00  24/02/2018                   828         Shipped   
1           10121  2765.90  07/05/2018                   757         Shipped   
2           10134  3884.34  01/07/2018                   703         Shipped   
3           10145  3746.70  25/08/2018                   649         Shipped   
4           10168  3479.76  28/10/2018                   586         Shipped   
...           ...      ...         ...                   ...             ...   
2742        10350  2244.40  02/12/2019                  2924         Shipped   
2743        10373  3978.51  31/01/2020                  2865         Shipped   
2744        10386  5417.57  01/03/2020                  2836     Not Shipped   
2745        10397  2116.16  28/03/2020                  2810         Shipped   
2746        10414  3079.44  06/05/2020                  2772     Not Shipped   

      productline  totalmsrp             customername        city  country  \
0     Motorcycles       2850        Land of Toys Inc.         NYC      USA   
1     Motorcycles       3230       Reims Collectables       Reims   France   
2     Motorcycles       3895          Lyon Souveniers       Paris   France   
3     Motorcycles       4275        Toys4GrownUps.com    Pasadena      USA   
4     Motorcycles       3420     Technics Stores Inc.  Burlingame      USA   
...           ...        ...                      ...         ...      ...   
2742        Ships       1080    Euro Shopping Channel      Madrid    Spain   
2743        Ships       1566  Oulu Toy Supplies, Inc.        Oulu  Finland   
2744        Ships       2322    Euro Shopping Channel      Madrid    Spain   
2745        Ships       1836             Alpha Cognac    Toulouse   France   
2746        Ships       2538        Gifts4AllAges.com      Boston      USA   

     productcode dealsize       contactname  \
0       S10_1678    Small           Kwai Yu   
1       S10_1678    Small      Paul Henriot   
2       S10_1678   Medium   Daniel Da Cunha   
3       S10_1678   Medium       Julie Young   
4       S10_1678   Medium       Juri Hirano   
...          ...      ...               ...   
2742    S72_3212    Small      Diego Freyre   
2743    S72_3212   Medium  Pirkko Koskitalo   
2744    S72_3212   Medium      Diego Freyre   
2745    S72_3212    Small    Annette Roulet   
2746    S72_3212   Medium      Juri Yoshido   

                                        completeaddress  
0              897 Long Airport Avenue, NYC, 10022, USA  
1              59 rue de l'Abbaye, Reims, 51100, France  
2     27 rue du Colonel Pierre Avia, Paris, 75508, F...  
3              78934 Hillside Dr., Pasadena, 90003, USA  
4             9408 Furth Circle, Burlingame, 94217, USA  
...                                                 ...  
2742           C/ Moralzarzal, 86, Madrid, 28034, Spain  
2743                  Torikatu 38, Oulu, 90110, Finland  
2744           C/ Moralzarzal, 86, Madrid, 28034, Spain  
2745     1 rue Alsace-Lorraine, Toulouse, 31000, France  
2746             8616 Spinnaker Dr., Boston, 51003, USA  

[2747 rows x 14 columns]

In [29]:
#  Basic cleaning (drop + date convert)


import pandas as pd
import numpy as np

# Copy original
df = data.copy()

# 1) days_since_lastorder drop 
if 'days_since_lastorder' in df.columns:
    df = df.drop(columns=['days_since_lastorder'])

# 2) orderdate to datetime conversion
df['orderdate'] = pd.to_datetime(df['orderdate'], dayfirst=True)

# 3) Analysis date for RFM
analysis_date = df['orderdate'].max() + pd.Timedelta(days=1)
print("Analysis date:", analysis_date.date())


Analysis date: 2020-06-01


In [30]:
# STEP 1 – Customer-level RFM + extra features

# Helper: mode safely lene ke liye
def safe_mode(x):
    m = x.mode()
    if m.empty:
        return np.nan
    return m.iloc[0]

# 👉 Customer-level aggregation
cust = (
    df
    .groupby('customername')
    .agg(
        recency_days = ('orderdate', lambda x: (analysis_date - x.max()).days),
        frequency_orders = ('ordernumber', 'nunique'),
        monetary_sales = ('sales', 'sum'),
        
        avg_totalmsrp = ('totalmsrp', 'mean'),
        n_unique_products = ('productcode', 'nunique'),
        
        # Shipping reliability = % of orders shipped
        shipping_reliability = ('shipping_status', lambda x: (x == 'Shipped').mean()),  # Did not understood this shipping reliability 1 how ?
        
        # Dominant attributes per customer
        main_country = ('country', safe_mode),
        main_city = ('city', safe_mode),
        main_productline = ('productline', safe_mode),
        main_dealsize = ('dealsize', safe_mode)
    )
    .reset_index()
)

print(cust.head())
print(cust.shape)


              customername  recency_days  frequency_orders  monetary_sales  \
0           AV Stores, Co.           197                 3       157807.81   
1             Alpha Cognac            65                 3        70488.44   
2       Amica Models & Co.           266                 2        94117.26   
3  Anna's Decorations, Ltd            84                 4       153996.13   
4        Atelier graphique           189                 3        24179.96   

   avg_totalmsrp  n_unique_products  shipping_reliability main_country  \
0    3194.588235                 37                   1.0           UK   
1    3357.450000                 20                   1.0       France   
2    3511.230769                 26                   1.0        Italy   
3    3325.000000                 41                   1.0    Australia   
4    3572.000000                  7                   1.0       France   

      main_city main_productline main_dealsize  
0    Manchester     Vintage Cars     

In [31]:
cust

customername  recency_days  frequency_orders  \
0                AV Stores, Co.           197                 3   
1                  Alpha Cognac            65                 3   
2            Amica Models & Co.           266                 2   
3       Anna's Decorations, Ltd            84                 4   
4             Atelier graphique           189                 3   
..                          ...           ...               ...   
84              Vida Sport, Ltd           276                 2   
85              Vitachrome Inc.           209                 3   
86     Volvo Model Replicas, Co           195                 4   
87  West Coast Collectables Co.           489                 2   
88            giftsbymail.co.uk           213                 2   

    monetary_sales  avg_totalmsrp  n_unique_products  shipping_reliability  \
0        157807.81    3194.588235                 37              1.000000   
1         70488.44    3357.450000                 20              1.000000   
2         94117.26    3511.230769                 26              1.000000   
3        153996.13    3325.000000                 41              1.000000   
4         24179.96    3572.000000                  7              1.000000   
..             ...            ...                ...                   ...   
84       117713.56    3931.677419                 17              1.000000   
85        88041.26    3282.560000                 25              1.000000   
86        75754.88    3862.473684                 19              0.684211   
87        46084.64    3699.461538                 13              1.000000   
88        78240.84    3061.923077                 26              1.000000   

   main_country     main_city main_productline main_dealsize  
0            UK    Manchester     Vintage Cars        Medium  
1        France      Toulouse            Ships        Medium  
2         Italy        Torino     Vintage Cars         Small  
3     Australia  North Sydney     Classic Cars         Small  
4        France        Nantes     Classic Cars        Medium  
..          ...           ...              ...           ...  
84  Switzerland        Gensve     Classic Cars        Medium  
85          USA           NYC      Motorcycles         Small  
86       Sweden          Lule     Classic Cars        Medium  
87          USA       Burbank     Vintage Cars        Medium  
88           UK         Cowes           Planes         Small  

[89 rows x 11 columns]

In [32]:
cust.to_csv("C:/Users/HARDEEP/OneDrive/Docs/Automobile-Sales-Logistics/Data/Cleaned-Data/cust_output.csv", index=False)

In [33]:
import os
os.getcwd()


'C:\\Users\\HARDEEP'